### PR TITLE
CGT-2540: Controller actions for confirming the correspondence address

### PR DESCRIPTION
**Pull request by macgrewal**

### CODE REVIEW

File: app/forms/YesNoForm.scala [hunk 1]
```diff
@@ -34,13 +37,14 @@ class YesNoForm @Inject()(val messagesApi: MessagesApi) extends I18nSupport {
...
   val yesNoForm = Form(
-    mapping (
+    mapping(
       "response" -> text
         .verifying(Messages("errors.mandatory"), nonEmptyCheck)
         .transform(textToBoolean, booleanToText)
     )(YesNoModel.apply)(YesNoModel.unapply)
   )
+
+  def validate[R](hasErrors: Form[YesNoModel] => R, success: YesNoModel => R)(implicit request: Request[_]): R =
```
Comment (miranda-hawkes) [inline app/forms/YesNoForm.scala]: What does R stand for here? Would T not be better as it's more frequently used

File: app/controllers/CorrespondenceAddressConfirmController.scala [hunk 2]
```diff
@@ -16,23 +16,79 @@
...
+import play.api.mvc.{Action, AnyContent, Result}
 import uk.gov.hmrc.play.frontend.controller.FrontendController
 
 import scala.concurrent.Future
 
 @Singleton
-class CorrespondenceAddressConfirmController @Inject()(appConfig: AppConfig, keystoreConnector: KeystoreConnector) extends FrontendController {
+class CorrespondenceAddressConfirmController @Inject()(appConfig: AppConfig,
+                                                       val messagesApi: MessagesApi,
+                                                       stateService: KeystoreConnector,
```
Comment (miranda-hawkes) [inline app/controllers/CorrespondenceAddressConfirmController.scala]: Would sessionService be a bit more meaningful? Not sure on this one

File: test/controllers/CorrespondenceAddressConfirmControllerSpec.scala [hunk 3]
```diff
@@ -16,16 +16,245 @@
...
+
+      "return a return a 200 response code" in {
+        status(result) shouldBe 200
+      }
+
+      "return the correct view" in {
+        document.title shouldBe messages.title
+      }
+    }
+
```
Comment (miranda-hawkes) [inline test/controllers/CorrespondenceAddressConfirmControllerSpec.scala]: Can you add a test for when a user is unauthorised - redirect to the dummy-unauthorised-url? And one for the submit action please

File: app/forms/YesNoForm.scala (hunk 1 above)
Comment (macgrewal) [inline app/forms/YesNoForm.scala]: R corresponds to the `bindFromRequest.fold()` function I've abstracted away to make the code int he controller simpler. It stands for Result as both success or failure should return a result.

File: app/controllers/CorrespondenceAddressConfirmController.scala [hunk 4]
```diff
@@ -34,7 +34,7 @@ import scala.concurrent.Future
 @Singleton
 class CorrespondenceAddressConfirmController @Inject()(appConfig: AppConfig,
                                                        val messagesApi: MessagesApi,
-                                                       stateService: KeystoreConnector,
+                                                       sessionervice: KeystoreConnector,
```
Comment (miranda-hawkes) [inline app/controllers/CorrespondenceAddressConfirmController.scala]: Sorry Mac, typo here

**Pull request merged**